src/core/layertree/qgslayertreemodellegendnode.cpp

```diff
@@ -604,7 +619,94 @@ void QgsSymbolLegendNode::updateLabel()
   emit dataChanged();
 }
 
+QString QgsSymbolLegendNode::evaluateLabel( QgsExpressionContext context, QString label )
+{
+  if ( !mLayerNode )
+    return QString();
+
+  QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( mLayerNode->layer() );
+
+  if ( vl && label.isEmpty() )
+  {
+    mLabel = getCurrentLabel();
+    if ( ! mLayerNode->expression().isEmpty() )
+      mLabel = evaluateLabelExpression( "[%" + mLayerNode->expression() + "%]", vl, context );
+    else if ( mLabel.contains( "[%" ) )
+      mLabel = evaluateLabelExpression( mLabel, vl, context );
+
+    emit dataChanged();
+    return mLabel;
+  }
+  else if ( vl )
+  {
+    if ( ! mLayerNode->expression().isEmpty() )
+      label = evaluateLabelExpression( label + "[%" + mLayerNode->expression() + "%]", vl, context );
+    else if ( label.contains( "[%" ) )
+      label = evaluateLabelExpression( label, vl, context );
+    return label;
+  }
+  else
+  {
+    if ( label.isEmpty() )
+      return mLabel;
+    else
+      return label;
+  }
+}
+
+QgsExpressionContext QgsSymbolLegendNode::createExpressionContext( QgsExpressionContext context ) const
+{
```

Comment:
Create the proper context and append it to the provided context, this is important to get the layout context to the evaluation).